### PR TITLE
non-contiguous and structured numpy arrays

### PR DESCRIPTION
**Pull request by fschlimb**
* handling non-contiguous numpy arrays without copying the entire array
* handling structured (heterogenous) numpy arrays

Somewhat relaxing #4 

@oleksandr-pavlyk @triskadecaepyon @Alexander-Makaryev

### CODE REVIEW

File: src/npy4daal.h [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+            // iterate through column, use casting functions to upcast, dataptr will point to current element
+            void ** dataptr = (void **) NpyIter_GetDataPtrArray(iter);
+
+            PyGILState_Release(__state);
+
+            if(WBack) { // could be templeate arg to eliminate conditional, prefer smaller binaires for now
+                do {
+                    daal::data_management::data_feature_utils::getVectorDownCast(
+                        f.indexType,
+                        daal::data_management::data_feature_utils::getInternalNumType<T>())(1,
```
[the remark this answers is not in the document]
Comment (RuslanIsrafilov) [reply src/npy4daal.h]: Do you always need to cast the data and copy it to `block`? is there situation when you can get the numpy pointer and pass it to block (via `block::setPtr`) directly without copying?

File: src/npy4daal.h [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+
+    virtual daal::services::Status resize(size_t nrows) DAAL_C11_OVERRIDE
+    {
+        std::cerr << "Resizing numpy array through daal not supported." << std::endl;
+        return daal::services::Status(daal::services::ErrorMethodNotSupported);
+    }
+
+    virtual int getSerializationTag() const
+    {
+        return daal::SERIALIZATION_AOS_NT_ID;
```
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: I thought so, too. We used it that way in pydaal, not sure if it was ever tested.
Can I just use any value as long as it does not conflict with existing tags?

File: src/npy4daal.h [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    template<typename T, bool WBack>
+    static void do_cpy(PyArrayObject * ary, daal::data_management::NumericTableDictionaryPtr & ddict,
+                       daal::data_management::BlockDescriptor<T>& block, size_t startcol, size_t ncols, size_t startrow, size_t nrows)
+    {
+        // Handle zero-sized arrays specially
+        if (PyArray_SIZE(ary) == 0) {
+            return;
+        }
+
+        auto __state = PyGILState_Ensure();
```
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: yes

File: src/npy4daal.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: This class handles 2 cases: AOS and non-contiguous arrays. Homogen, contiguous arrays are handled separately (elsewhere) by passing the data pointer to a HomogenNumericTable. For these 2 detecting if setting the pointer is possible is probably too costly for a probably super rare corner case.

File: src/npy4daal.h [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+#if PY_MAJOR_VERSION < 3
+        char * ds = NULL;
+        PyString_AsStringAndSize(PyObject_Repr((PyObject*)PyArray_DESCR(ary)), &ds, &len);
+#else
+        const char * ds = PyUnicode_AsUTF8AndSize(PyObject_Repr((PyObject*)PyArray_DESCR(ary)), &len);
+#endif
+        if(ds == NULL) {
+            this->_status.add(daal::services::UnknownError);
+            PyGILState_Release(__state);
+            return daal::services::Status();
```
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: like ```return daal::services::Status(daal::services::UnknownError)```?

**Comment by fschlimb** [comment]
@RuslanIsrafilov When is de/serialization actually needed? The user/programmer will never directly use it because it's only an internal class that is not exposed. I guess DAAL will also never use this class (e.g. it will never create a new object of this class). So in which situations will DAAL call de/serialize() on a NumericTable?

File: src/npy4daal.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: fixed

File: src/npy4daal.h [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+        // The location of the data pointer which the iterator may update
+        char ** dataptr = NpyIter_GetDataPtrArray(iter);
+        // The location of the stride which the iterator may update
+        npy_intp * strideptr = NpyIter_GetInnerStrideArray(iter);
+        // The location of the inner loop size which the iterator may update
+        npy_intp * innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+        if(NpyIter_GetDescrArray(iter)[0]->elsize != sizeof(T)) {
+            NpyIter_Deallocate(iter);
+            throw std::invalid_argument("Encountered unexpected element size or type when copying block.");
```
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: fixed

File: src/npy4daal.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: done

File: src/npy4daal.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (fschlimb) [reply src/npy4daal.h]: Moved to throwing exceptions.

**Comment by RuslanIsrafilov** [comment]
> When is de/serialization actually needed? 
De/serialization methods were implemented, so I thought it was needed in daal4py for distributed execution or pickling. No problems with serialization if it's called directly via `NpyNumericTable::de/serialize(...)`.

> So in which situations will DAAL call de/serialize() on a NumericTable?
Almost all results and models are collection of NTs, so if you serialize result/model, it causes nested serialization of NT. DAAL API enables you to set you own derived NT to the result via `Result::set(...)`. Obviously, with `NpyNumericTable` deserialization would not work in this scenario. 

If this table is going to be used ONLY as an input, this implementation is OK. But if at some point, somebody will use this type of table as a nested in serializable object (e.g. result/model), it would create issue that is hard to debug. 

Anyway, serialization tag should be changed.
> Can I just use any value as long as it does not conflict with existing tags?
Yes.

**Comment by fschlimb** [comment]
Ok, @RuslanIsrafilov thanks for the review! I added custom serialization tag.
If can send me instructions/code how to harden the serialization issue I can add it later.

**Pull request merged**